### DISCUSSION TITLE
How to use signed URL for kinesis-video-signaling get-ice-server-config API?

### DISCUSSION THREAD

**Discussion by overheat** [awslabs/aws-sdk-rust#1160] [Q&A] [answered]
I need a rust version function for this typescript fuction:
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/blob/de209884fcd388c3fea4937a5b99354ef9519af3/src/SigV4RequestSigner.ts#L43

How to implemented such code by aws-sdk-rust?

**Comment by Velfi** [comment]
We don't have a Rust equivalent of the "Amazon Kinesis Video Streams with WebRTC" high-level library. You can submit a feature request for it but I don't expect we'll get to it any time soon.

In the meantime, if you'd like to take a look at our sigv4 signing code, [it can be found here](https://github.com/smithy-lang/smithy-rs/blob/main/aws/rust-runtime/aws-sigv4/src/http_request/sign.rs#L472).

**Reply by ysaito1001** [reply, marked as answer]
Admittedly, the signing API is not ergonomic today, but nevertheless you can also find an example [here](https://docs.rs/aws-sigv4/1.2.1/aws_sigv4/http_request/index.html#example-signing-an-http-request). Since the link above shows a function that creates a URL with query params, you probably want to have a [SigningSettings](https://docs.rs/aws-sigv4/1.2.1/aws_sigv4/http_request/struct.SigningSettings.html) with a field `signature_location` set to `SignatureLocation::QueryParams`.

**Reply by overheat** [reply]
@ysaito1001 
Yes, now I still need 2 more item `X-Amz-ChannelARN` and `X-Amz-ClientId` in the signed URL.
Do you know how to add them manually?

**Reply by ysaito1001** [reply]
Referring back to [the example](https://docs.rs/aws-sigv4/1.2.1/aws_sigv4/http_request/index.html#example-signing-an-http-request), after the end of the last line
```
signing_instructions.apply_to_request_http1x(&mut my_req);
```
maybe you can call `.uri_mut` on `my_req` to include `X-Amz-ChannelARN` and `X-Amz-ClientId` (assuming their values do not need to be part of the signing process and can be added to the signed URL) ?

**Reply by overheat** [reply]
@ysaito1001 right!
only one thing that for `ChannelARN` there are some characters like `:` need to be translate to `%3`. Any convenient method suggestion?

**Reply by ysaito1001** [reply]
I'd give [fmt_string](https://docs.rs/aws-smithy-http/latest/aws_smithy_http/query/fn.fmt_string.html) a try ([example usage](https://github.com/smithy-lang/smithy-rs/blob/4176e8011e149a2ef0914309a54d6ab1aa7e50a5/rust-runtime/aws-smithy-http/src/query.rs#L79)).

**Reply by dadleyy** [reply]
:wave: bopping into this discussion - its the first kvs related discussion I could find for the rust sdk - I've been looking to a pure rust implementation of an application that streams video data to KVS for personal projects and created a minimal, functioning example here - https://github.com/dadleyy/rs-kvs-streamer.

the generation of the signed requests happens in [`kvs.rs`](https://github.com/dadleyy/rs-kvs-streamer/blob/28419a0aa77ac3f7f3b7a955998771b15b97aad4/src/kvs.rs#L109)

the other part that was really tricky was getting the video data into the mkv+h264 format expected by the `putMedia` api; I ended up combining the `openh264` + `webm_iterable` crates to get the job done. it might make sense to wrap that logic in a community owned crate in-and-of-itself.
